cmd/clusterctl/client/cluster/objectgraph.go

```diff
@@ -674,3 +796,68 @@ func (o *objectGraph) checkVirtualNode() {
 		}
 	}
 }
+
+// Ensure objects which are referenced across namespaces are not deleted.
+func (o *objectGraph) setShouldNotDelete(ctx context.Context, namespace string) error {
+	if namespace == "" {
+		return nil
+	}
+
+	// If there are ClusterClasses outside of the namespace being moved, those CC should not be moved (we are moving a namespace).
+	for _, class := range o.getClusterClasses() {
+		if class.identity.Namespace != namespace {
+			class.shouldNotDelete = true
+			// Ensure that also the templates referenced by the CC won't be moved.
+			o.setShouldNotDeleteHierarchy(class)
+		}
+	}
+
+	// If there are clusters outside of the namespace being moved and referencing one of Class in the namespace being moved,
+	// ensure they are not moved.
+	discoveryBackoff := newReadBackoff()
+	allClusters := &clusterv1.ClusterList{}
+	if err := retryWithExponentialBackoff(ctx, discoveryBackoff, func(ctx context.Context) error {
+		return getObjList(ctx, o.proxy, nil, []client.ListOption{}, allClusters)
+	}); err != nil {
+		return err
+	}
+
+	for _, cluster := range allClusters.Items {
+		// ignore cluster in the namespace being moved.
+		if cluster.Namespace == namespace {
+			continue
+		}
+
+		// ignore cluster not using a CC.
+		if cluster.Spec.Topology == nil {
+			continue
+		}
+
+		// ignore cluster not referencing a CC in the namespace being moved.
+		if cluster.Spec.Topology.ClassNamespace != namespace {
+			continue
+		}
+
+		// Otherwise mark the reference CC as should not be moved.
+		for _, class := range o.getClusterClasses() {
+			if class.identity.Namespace == cluster.Spec.Topology.ClassNamespace && class.identity.Name == cluster.Spec.Topology.Class {
+				class.shouldNotDelete = true
+				// Ensure that also the templates referenced by the CC won't be moved.
+				o.setShouldNotDeleteHierarchy(class)
+			}
+		}
+	}
+
+	return nil
+}
+
+// setShouldNotDeleteHierarchy sets should not delete for a node and for its own dependents/softDependents.
+func (o *objectGraph) setShouldNotDeleteHierarchy(node *node) {
+	node.shouldNotDeleteHierarchy = true
```

Comment:
I think we can further simplify this and drop entirely node.shouldNotDeleteHierarchy (we do not need to distinguish between shouldNotDelete and shouldNotDeleteHierarchy)
```suggestion
	node.shouldNotDelete = true
```